src/core/types.ts

```diff
@@ -676,10 +676,10 @@ export type MutationObserverResult<
   | MutationObserverErrorResult<TData, TError, TVariables, TContext>
   | MutationObserverSuccessResult<TData, TError, TVariables, TContext>
 
-export interface QueryClientConfig {
+export interface QueryClientConfig<TError = unknown> {
   queryCache?: QueryCache
   mutationCache?: MutationCache
-  defaultOptions?: DefaultOptions
+  defaultOptions?: DefaultOptions<TError>
```

Comment:
I don't think this works. The `config` passed to the `QueryClient` constructor is still not generic. Additionally, the caches also have global onError callbacks that wouldn't get this generic.

all in all, I think its simpler to go with the runtime check.